File: src/simupoke/bench.py
"""Benchmarks & optimiseur de SP (§11.1 — « seuils de survie / de KO »).

Brique de base réutilisée par B1/B3 : au lieu de juste calculer des dégâts pour
un build donné, on répond aux questions **inverses** que se pose un joueur VGC :

  - *speed tiers* : qui dépasse qui en vitesse (avec Scarf, Tailwind, paralysie,
    Trick Room) ?
  - **combien de SP en Vitesse** pour dépasser tel adversaire ?
  - **combien de SP défensifs** (PV + Déf/Déf.Spé) pour **survivre** à telle
    attaque ?
  - **combien de SP offensifs** (Atq/Atq.Spé) pour **garantir un KO** en N coups ?

Tout s'appuie sur le moteur figé (`stats` + `damage`) — aucune donnée réseau,
une seule source de vérité. Le modèle de SP Champions (§8.3) rend l'espace de
recherche trivial : chaque stat est un entier dans [0, 32], on énumère.

Note budget : ces optimiseurs raisonnent **stat par stat** (le SP minimal requis
dans la/les stat(s) concernée(s)). La contrainte de budget global (66 SP au
total, §8.3) reste à la charge du joueur qui assemble son spread — l'outil dit
« il te faut au moins X SP ici », pas « voici ton spread complet ».

Focus Sash / Fermeté (§11.1) sont pris en compte : un défenseur à PV pleins
**encaisse** un coup fatal (reste 1 PV), donc la survie est garantie sans SP et
un OHKO devient un 2HKO minimum.
"""

from __future__ import annotations

from dataclasses import dataclass, replace

from .stats import SP_CAP_PER_STAT
from .basestats import to_id
from .moves import get_move
from .model import PokemonState, FieldState
from .damage import calculate
from .combat import effective_speed


# ---------------------------------------------------------------------------
# Vitesse effective (avec Tailwind, en plus de Scarf/paralysie/boosts)
# ---------------------------------------------------------------------------

def compute_speed(state: PokemonState, *, tailwind: bool = False) -> int:
    """Vitesse effective d'un Pokémon, Tailwind inclus.

    Reprend `combat.effective_speed` (base + boosts + Choice Scarf + paralysie)
    et applique le doublement de Tailwind par-dessus (comme Showdown).
    """
    spe = effective_speed(state)
    if tailwind:
        spe *= 2
    return spe


@dataclass
class SpeedEntry:
    species: str
    speed: int
    notes: list[str]              # ex. ["Choice Scarf", "Tailwind", "paralysé"]


def _speed_notes(state: PokemonState, tailwind: bool) -> list[str]:
    notes: list[str] = []
    if to_id(state.item or "") == "choicescarf":
        notes.append("Choice Scarf")
    if tailwind:
        notes.append("Tailwind")
    if state.status == "par":
        notes.append("paralysé")
    boost = state.boosts.get("spe", 0)
    if boost:
        notes.append(f"Vit {boost:+d}")
    return notes


def speed_tiers(states: list[PokemonState], *,
                tailwinds: list[bool] | None = None,
                trick_room: bool = False) -> list[SpeedEntry]:
    """Classe des Pokémon par vitesse effective.

    `tailwinds` : liste parallèle de drapeaux Tailwind (défaut : tous False).
    `trick_room` : si vrai, l'ordre d'ACTION s'inverse (le plus lent agit en
    premier) — on trie alors en vitesse croissante. Le champ `speed` reste la
    vraie vitesse ; seul l'ordre de la liste change.
    """
    tw = tailwinds if tailwinds is not None else [False] * len(states)
    entries = [
        SpeedEntry(species=s.species, speed=compute_speed(s, tailwind=t),
                   notes=_speed_notes(s, t))
        for s, t in zip(states, tw)
    ]
    entries.sort(key=lambda e: e.speed, reverse=not trick_room)
    return entries


# ---------------------------------------------------------------------------
# Optimiseur : Vitesse pour dépasser une cible
# ---------------------------------------------------------------------------

@dataclass
class OutspeedResult:
    feasible: bool
    sp: int | None                # SP minimal en Vitesse (None si impossible)
    my_speed: int | None          # ma vitesse au SP trouvé (ou au max testé)
    target_speed: int
    ties_only: bool = False       # au mieux on égalise (départage 50/50)

    def line(self) -> str:
        if self.feasible:
            return (f"Vitesse : {self.sp} SP suffisent "
                    f"({self.my_speed} vs {self.target_speed}).")
        if self.ties_only:
            return (f"Vitesse : impossible de dépasser ; on peut au mieux "
                    f"ÉGALISER à {self.target_speed} (départage 50/50).")
        return (f"Vitesse : hors de portée — même à {SP_CAP_PER_STAT} SP, "
                f"{self.my_speed} < {self.target_speed}.")


def _with_sp(state: PokemonState, key: str, value: int) -> PokemonState:
    return replace(state, stat_points={**state.stat_points, key: value})


def _endures(state: PokemonState) -> bool:
    """Focus Sash / Fermeté : survit à UN coup s'il part des PV pleins."""
    if state.current_hp_pct < 1.0:
        return False
    return (to_id(state.item or "") == "focussash"
            or to_id(state.ability or "") == "sturdy")


def _effective_ko_hits(raw: int | None, endures: bool) -> int | None:
    """Nb de coups pour KO garanti, en tenant compte de Focus Sash / Fermeté.

    Un coup fatal partant des PV pleins est encaissé (reste 1 PV) → il faut un
    coup de plus qu'un OHKO « brut ».
    """
    if raw is None:
        return None
    return 2 if (endures and raw == 1) else raw


def min_sp_to_outspeed(me: PokemonState, target: PokemonState,
                       *, me_tailwind: bool = False,
                       target_tailwind: bool = False,
                       strict: bool = True) -> OutspeedResult:
    """SP minimal en Vitesse pour (dé)passer `target`.

    `strict=True` : il faut être **strictement** plus rapide (départage évité).
    `strict=False` : égaliser suffit.
    """
    target_speed = compute_speed(target, tailwind=target_tailwind)
    best_speed = -1
    ties = False
    for sp in range(0, SP_CAP_PER_STAT + 1):
        s = compute_speed(_with_sp(me, "spe", sp), tailwind=me_tailwind)
        best_speed = s
        if s > target_speed or (not strict and s >= target_speed):
            return OutspeedResult(True, sp, s, target_speed,
                                  ties_only=(s == target_speed))
        if s == target_speed:
            ties = True
    return OutspeedResult(False, None, best_speed, target_speed, ties_only=ties)


# ---------------------------------------------------------------------------
# Optimiseur : survie défensive
# ---------------------------------------------------------------------------

@dataclass
class SurviveResult:
    feasible: bool
    hp_sp: int | None
    def_sp: int | None
    total_sp: int | None
    move: str
    stat: str                     # "def" ou "spd" (côté défensif sollicité)
    max_pct: float | None = None  # % PV du roll haut au spread trouvé
    by_endure: bool = False       # survie garantie par Focus Sash / Fermeté

    def line(self) -> str:
        if self.by_endure:
            return (f"Survie à {self.move} : garantie par Focus Sash / Fermeté "
                    f"(depuis PV pleins) — 0 SP requis.")
        if not self.feasible:
            return (f"Survie à {self.move} : impossible même à fond "
                    f"(PV + {self.stat}).")
        return (f"Survie à {self.move} : {self.total_sp} SP "
                f"(PV {self.hp_sp} / {self.stat} {self.def_sp}) — "
                f"le roll haut fait {self.max_pct:.0f}% max.")


def min_sp_to_survive(defender: PokemonState, attacker: PokemonState,
                      move: str, field: FieldState | None = None, *,
                      crit: bool = False) -> SurviveResult:
    """SP défensif total minimal (PV + Déf/Déf.Spé) pour survivre au roll haut.

    On énumère les répartitions par total croissant et on renvoie la première
    (donc de coût minimal) qui garantit la survie même sur le roll max. Le
    défenseur est évalué à PLEINS PV (un seuil de survie « propre »).
    """
    m = get_move(move)
    stat = "def" if m.is_physical else "spd"
    base = replace(defender, current_hp_pct=1.0)
    # Focus Sash / Fermeté : survie garantie depuis PV pleins, sans SP.
    if _endures(base):
        return SurviveResult(True, 0, 0, 0, m.name, stat, by_endure=True)
    cap = SP_CAP_PER_STAT
    for total in range(0, 2 * cap + 1):
        lo = max(0, total - cap)
        hi = min(cap, total)
        for hp_sp in range(lo, hi + 1):
            def_sp = total - hp_sp
            trial = replace(base, stat_points={
                **base.stat_points, "hp": hp_sp, stat: def_sp})
            r = calculate(attacker, trial, m, field, crit=crit)
            if r.max_damage < r.defender_max_hp:
                return SurviveResult(True, hp_sp, def_sp, total, m.name, stat,
                                     max_pct=r.max_pct)
    return SurviveResult(False, None, None, None, m.name, stat)


# ---------------------------------------------------------------------------
# Optimiseur : SP offensif pour garantir un KO
# ---------------------------------------------------------------------------

@dataclass
class KoResult:
    feasible: bool
    sp: int | None
    move: str
    stat: str                     # "atk" ou "spa"
    hits: int                     # KO recherché en N coups
    min_pct: float | None = None  # % PV du roll bas au SP trouvé
    blocked_by_endure: bool = False   # empêché par Focus Sash / Fermeté

    def line(self) -> str:
        what = "OHKO" if self.hits == 1 else f"KO en {self.hits}"
        if self.blocked_by_endure:
            return (f"{what} avec {self.move} : bloqué par Focus Sash / Fermeté "
                    f"(cible à PV pleins) — impossible en {self.hits}.")
        if not self.feasible:
            return (f"{what} avec {self.move} : hors de portée même à "
                    f"{SP_CAP_PER_STAT} SP {self.stat}.")
        return (f"{what} avec {self.move} : {self.sp} SP {self.stat} "
                f"(roll bas {self.min_pct:.0f}%).")


def min_sp_to_ko(attacker: PokemonState, defender: PokemonState,
                 move: str, field: FieldState | None = None, *,
                 hits: int = 1, crit: bool = False) -> KoResult:
    """SP offensif minimal (Atq ou Atq.Spé) pour un KO **garanti** en `hits`.

    « Garanti » = même le roll bas tue en `hits` coups. Le défenseur est pris
    dans l'état fourni (PV courants inclus), ce qui permet aussi de raisonner
    « combien pour finir un adversaire déjà entamé ».
    """
    m = get_move(move)
    stat = "atk" if m.is_physical else "spa"
    endures = _endures(defender)
    for sp in range(0, SP_CAP_PER_STAT + 1):
        trial = _with_sp(attacker, stat, sp)
        r = calculate(trial, defender, m, field, crit=crit)
        g = _effective_ko_hits(r.guaranteed_ko_hits, endures)
        if g is not None and g <= hits:
            return KoResult(True, sp, m.name, stat, hits, min_pct=r.min_pct)
    # KO en `hits` inatteignable : signaler si c'est Focus Sash / Fermeté qui
    # l'empêche (un OHKO devient un 2HKO minimum).
    blocked = endures and hits < 2
    return KoResult(False, None, m.name, stat, hits, blocked_by_endure=blocked)

File: web/bench.js
/* SimuPoke — Seuils & optimiseur de SP (port JS de src/simupoke/bench.py).
 *
 * Construit au-dessus du moteur (engine.js) : mêmes réponses que le cœur Python
 * (vérifié par parité). Utilisable navigateur (window.SimuBench) et Node.
 */
(function (root, factory) {
  const mod = factory();
  if (typeof module !== 'undefined' && module.exports) module.exports = mod;
  else root.SimuBench = mod;
})(typeof self !== 'undefined' ? self : this, function () {
  'use strict';

  const CAP = 32;
  const toId = (s) => (s || '').toLowerCase().replace(/[^a-z0-9]/g, '');
  const boostMult = (st) => (st >= 0 ? (2 + st) / 2 : 2 / (2 - st));

  function makeBench(engine) {
    // --- Vitesse effective (base + boosts + Scarf + paralysie), Tailwind en option
    function effectiveSpeed(state) {
      let spe = engine.battleStats(state).spe;
      spe = Math.floor(spe * boostMult((state.boosts && state.boosts.spe) || 0));
      if (toId(state.item) === 'choicescarf') spe = Math.floor(spe * 1.5);
      if (state.status === 'par') spe = Math.floor(spe * 0.5);
      return spe;
    }
    function computeSpeed(state, tailwind) {
      const s = effectiveSpeed(state);
      return tailwind ? s * 2 : s;
    }
    const withSp = (state, key, v) =>
      Object.assign({}, state, { sp: Object.assign({}, state.sp, { [key]: v }) });

    function speedNotes(state, tailwind) {
      const n = [];
      if (toId(state.item) === 'choicescarf') n.push('Choice Scarf');
      if (tailwind) n.push('Tailwind');
      if (state.status === 'par') n.push('paralysé');
      const b = (state.boosts && state.boosts.spe) || 0;
      if (b) n.push('Vit ' + (b > 0 ? '+' : '') + b);
      return n;
    }
    function speedTiers(states, opts) {
      opts = opts || {};
      const tw = opts.tailwinds || states.map(() => false);
      const trickRoom = !!opts.trickRoom;
      const rows = states.map((s, i) => ({
        species: s.species, speed: computeSpeed(s, tw[i]), notes: speedNotes(s, tw[i]),
      }));
      rows.sort((a, b) => (trickRoom ? a.speed - b.speed : b.speed - a.speed));
      return rows;
    }

    function minSpToOutspeed(me, target, opts) {
      opts = opts || {};
      const strict = opts.strict !== false;
      const targetSpeed = computeSpeed(target, opts.targetTailwind);
      let best = -1, ties = false;
      for (let sp = 0; sp <= CAP; sp++) {
        const s = computeSpeed(withSp(me, 'spe', sp), opts.meTailwind);
        best = s;
        if (s > targetSpeed || (!strict && s >= targetSpeed)) {
          return { feasible: true, sp, mySpeed: s, targetSpeed, tiesOnly: s === targetSpeed };
        }
        if (s === targetSpeed) ties = true;
      }
      return { feasible: false, sp: null, mySpeed: best, targetSpeed, tiesOnly: ties };
    }

    function endures(state) {
      const full = (state.hpPct === undefined ? 1 : state.hpPct) >= 1;
      return full && (toId(state.item) === 'focussash' || toId(state.ability) === 'sturdy');
    }

    function minSpToSurvive(defender, attacker, move, field, opts) {
      opts = opts || {};
      const m = engine.move(move);
      const stat = m.category === 'Physical' ? 'def' : 'spd';
      const base = Object.assign({}, defender, { hpPct: 1 });
      if (endures(base)) return { feasible: true, hpSp: 0, defSp: 0, totalSp: 0, stat, byEndure: true, move: m.name };
      for (let total = 0; total <= 2 * CAP; total++) {
        const lo = Math.max(0, total - CAP), hi = Math.min(CAP, total);
        for (let hp = lo; hp <= hi; hp++) {
          const dsp = total - hp;
          const trial = Object.assign({}, base, { sp: Object.assign({}, base.sp, { hp, [stat]: dsp }) });
          const r = engine.calculate(attacker, trial, move, field, { crit: !!opts.crit, screen: opts.screen });
          if (r.max < r.maxHp) return { feasible: true, hpSp: hp, defSp: dsp, totalSp: total, stat, maxPct: r.maxPct, move: m.name };
        }
      }
      return { feasible: false, stat, move: m.name };
    }

    function effectiveKoHits(raw, end) {
      if (raw == null) return null;
      return (end && raw === 1) ? 2 : raw;
    }
    function minSpToKo(attacker, defender, move, field, opts) {
      opts = opts || {};
      const hits = opts.hits || 1;
      const m = engine.move(move);
      const stat = m.category === 'Physical' ? 'atk' : 'spa';
      const end = endures(defender);
      for (let sp = 0; sp <= CAP; sp++) {
        const trial = withSp(attacker, stat, sp);
        const r = engine.calculate(trial, defender, move, field, { crit: !!opts.crit });
        const g = effectiveKoHits(r.koGuaranteed, end);
        if (g != null && g <= hits) return { feasible: true, sp, stat, hits, minPct: r.minPct, move: m.name };
      }
      return { feasible: false, sp: null, stat, hits, blockedByEndure: end && hits < 2, move: m.name };
    }

    // --- Optimiseur de spread (port de optimize.py) ---
    function minDefGivenHp(meDef, obj, hp, defKey) {
      for (let dsp = 0; dsp <= CAP; dsp++) {
        const trial = Object.assign({}, meDef, { hpPct: 1, sp: Object.assign({}, meDef.sp, { hp, [defKey]: dsp }) });
        const r = engine.calculate(obj.attacker, trial, obj.move, obj.field || {}, { crit: !!obj.crit });
        if (r.max < r.maxHp) return dsp;
      }
      return null;
    }
    function solveDefense(meDef, survives) {
      const phys = survives.filter((o) => engine.move(o.move).category === 'Physical');
      const spec = survives.filter((o) => engine.move(o.move).category !== 'Physical');
      if (!phys.length && !spec.length) return { hp: 0, def: 0, spd: 0 };
      let best = null;
      for (let hp = 0; hp <= CAP; hp++) {
        let defNeed = 0, spdNeed = 0;
        for (const o of phys) defNeed = Math.max(defNeed, minDefGivenHp(meDef, o, hp, 'def') || 0);
        for (const o of spec) spdNeed = Math.max(spdNeed, minDefGivenHp(meDef, o, hp, 'spd') || 0);
        const total = hp + defNeed + spdNeed;
        if (best === null || total < best.total) best = { total, hp, def: defNeed, spd: spdNeed };
      }
      return best;
    }
    function optimizeSpread(species, nature, objectives, opts) {
      opts = opts || {};
      const budget = opts.budget || 66;
      const item = opts.item || null, ability = opts.ability || null;
      const me = (extra) => Object.assign({ species, nature, item, ability, sp: {} }, extra || {});
      const sp = { hp: 0, atk: 0, def: 0, spa: 0, spd: 0, spe: 0 };
      const unmet = [];
      const outs = objectives.filter((o) => o.kind === 'outspeed');
      const kos = objectives.filter((o) => o.kind === 'ko');
      const survives = objectives.filter((o) => o.kind === 'survive');

      for (const o of outs) {
        const r = minSpToOutspeed(me(), o.target, { strict: o.strict !== false, meTailwind: o.meTailwind, targetTailwind: o.targetTailwind });
        if (!r.feasible) unmet.push('dépasser ' + (o.label || o.target.species) + ' : hors de portée (vitesse)');
        else sp.spe = Math.max(sp.spe, r.sp);
      }
      for (const o of kos) {
        const r = minSpToKo(me(), o.defender, o.move, o.field || {}, { hits: o.hits || 1, crit: o.crit });
        const tag = o.label || (o.move + ' sur ' + o.defender.species);
        if (!r.feasible) unmet.push('KO ' + tag + ' : ' + (r.blockedByEndure ? 'bloqué par Focus Sash / Fermeté' : 'hors de portée (>' + CAP + ' SP ' + r.stat + ')'));
        else sp[r.stat] = Math.max(sp[r.stat], r.sp);
      }
      const meDef = me();
      const okSurv = [];
      for (const o of survives) {
        const chk = minSpToSurvive(meDef, o.attacker, o.move, o.field || {}, { crit: o.crit });
        if (!chk.feasible) unmet.push('survivre à ' + (o.label || (o.move + ' de ' + o.attacker.species)) + ' : hors de portée (même à fond)');
        else okSurv.push(o);
      }
      const d = solveDefense(meDef, okSurv);
      sp.hp = d.hp; sp.def = d.def; sp.spd = d.spd;

      const total = sp.hp + sp.atk + sp.def + sp.spa + sp.spd + sp.spe;
      if (total > budget) unmet.push('budget dépassé : ' + total + ' SP requis > ' + budget);
      let stats = null;
      try { stats = engine.battleStats({ species, nature, sp }); } catch (e) { stats = null; }
      return { feasible: unmet.length === 0, sp, total, budget, leftover: budget - total, stats, unmet };
    }

    return { effectiveSpeed, computeSpeed, speedTiers, minSpToOutspeed, minSpToSurvive, minSpToKo, optimizeSpread };
  }

  return { makeBench, toId };
});
